ambry-router/src/main/java/com/github/ambry/router/QuotaAwareOperationController.java

```diff
@@ -0,0 +1,193 @@
+    undeleteManager.poll(requestsToSend, requestsToDrop);
+  }
+
+  /**
+   * Add the specified {@link List} of {@link RequestInfo}s to the reqeust queue.
+   * @param requestInfos {@link List} of {@link RequestInfo} objects.
+   */
+  private void addToRequestQueue(List<RequestInfo> requestInfos) {
+    for (RequestInfo requestInfo : requestInfos) {
+      QuotaResource quotaResource = requestInfo.getChargeable().getQuotaResource();
+      if (quotaResource == null) {
+        quotaResource = UNKNOWN_QUOTA_RESOURCE;
+      }
+      requestQueue.putIfAbsent(quotaResource, new LinkedList<>());
+      requestQueue.get(quotaResource).add(requestInfo);
+    }
+  }
+
+  /**
+   * Drain the request queue based on quota and system resources and update the requests to be send to {@code requestsToSend}.
+   * @param requestsToSend a list of {@link RequestInfo} that will contain the requests to be sent out.
+   */
+  private void drainRequestQueue(List<RequestInfo> requestsToSend) {
+    drainQuotaCompliantRequests(requestsToSend);
+
+    List<QuotaResource> quotaResources = new ArrayList<>(requestQueue.keySet());
+    Collections.shuffle(quotaResources);
+    boolean done = false;
+    while (!requestQueue.isEmpty()) {
+      for (QuotaResource quotaResource : quotaResources) {
+        RequestInfo requestInfo = requestQueue.get(quotaResource).getFirst();
+        if (requestInfo.getChargeable().quotaExceedAllowed()) {
+          requestInfo.getChargeable().charge();
+          requestsToSend.add(requestInfo);
+          requestQueue.get(quotaResource).removeFirst();
+          if (requestQueue.get(quotaResource).isEmpty()) {
+            requestQueue.remove(quotaResource);
+          }
+        } else {
+          done = true;
+          break;
+        }
+      }
+      if (done) {
+        break;
+      }
+    }
+  }
+
+  /**
+   * Drain the request queue based on resource quota only and update the requests to be send to {@code requestsToSend}.
+   * @param requestsToSend a list of {@link RequestInfo} that will contain the requests to be sent out.
+   */
+  private void drainQuotaCompliantRequests(List<RequestInfo> requestsToSend) {
+    for (QuotaResource quotaResource : requestQueue.keySet()) {
+      if (quotaResource.equals(UNKNOWN_QUOTA_RESOURCE)) {
+        // If there are requests for which QuotaResource couldn't be found, this is most likely a bug.
+        // As a temporary hack, we will consider all those requests to be quota compliant.
+        requestsToSend.addAll(requestQueue.get(UNKNOWN_QUOTA_RESOURCE));
+        requestQueue.remove(UNKNOWN_QUOTA_RESOURCE);
+        continue;
+      }
+      while (!requestQueue.get(quotaResource).isEmpty()) {
+        RequestInfo requestInfo = requestQueue.get(quotaResource).getFirst();
+        if (requestInfo.getChargeable().check()) {
+          requestsToSend.add(requestInfo);
+          requestQueue.get(quotaResource).removeFirst();
+          requestInfo.getChargeable().charge();
```

Comment:
Technically check() & charge() should be atomic. You may want to change your interface to CheckAndCharge() and it charges in a thread-safe manner if check() returns true.

Reply:
You are right about this. I will fix this in the next iteration. The atomicity should actually come from the last layer of implementation which is `QuotaSource`. Will modify the call stack to reflect that.